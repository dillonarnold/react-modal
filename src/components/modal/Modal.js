import React, {Component} from 'react';
import {bool, func} from 'prop-types';
import ModalFooter from './ModalFooter';

/**
 * A popup modal component.
 */
class Modal extends Component {

  static propTypes = {
    /** Opens the modal. */
    isShown: bool,
    /** Called when the modal is closed by clicking on the backdrop, closing button, or pressing ESC. */
    onClose: func,
  };

  static defaultProps = {
    isShown: false,
    onClose: () => {}
  };

  constructor(props) {
    super(props);
    this.modalRef = React.createRef();
    this.closeButtonRef = React.createRef();
  }

  componentDidMount() {
    this.props.isShown && this.open();
  }

  componentDidUpdate(prevProps) {
    if (prevProps.isShown !== this.props.isShown) {
      this.props.isShown ? this.open() : this.close();
    }
  }

  componentWillUnmount() {
    this.close();
  }

  // Handle opening of modal
  open() {
    document.addEventListener('keydown', this.handleKeyDown);
    this.modalRef.current.classList.remove('modal--hide');
    this.modalRef.current.classList.add('modal--display');
    // Focus on first button for accessibility
    this.closeButtonRef.current.focus();
  }

  // Handle closing of modal
  close() {
    document.removeEventListener('keydown', this.handleKeyDown);
    this.modalRef.current.classList.remove('modal--display');
    this.modalRef.current.classList.add('modal--hide');
    this.props.onClose();
  }

  handleKeyDown = e => {
    (e.key === 'Escape' || e.keyCode === 27) && this.close();
  };

  handleBackgroundClick = e => {
    // Prevent event bubbling
    if (e.target === this.modalRef.current) {
      this.close();
    }
  };

  render() {
    const {children} = this.props;

    return (
      <div className="modal-background modal--hide" ref={this.modalRef} onClick={this.handleBackgroundClick}>
        <div className="modal" role="dialog">
          <div className="content">
            {children}
          </div>
          <ModalFooter onClose={() => this.close()} closeButtonRef={this.closeButtonRef}/>
        </div>
      </div>
    );
  };
}

export default Modal;
