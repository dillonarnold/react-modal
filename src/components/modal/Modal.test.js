import React from 'react';
import { mount } from 'enzyme';
import toJson from 'enzyme-to-json';
import Modal from './Modal';

describe('Modal component', () => {

  it('should render properly', () => {
    const component = mount(
      <Modal
        isShown
      />,
    );

    expect(toJson(component)).toMatchSnapshot();

  });

  it('closes only on clicking outside of modal', () => {
    const callback = jest.fn();
    const component = mount(
      <Modal
        isShown
        onClose={callback}
      />,
    );

    // Simulate click inside
    component.find('div.modal').simulate('click');
    expect(callback).not.toBeCalled();

    // Simulate click outside
    component.find('div.modal-background').simulate('click');
    expect(callback).toBeCalled();
  });
});