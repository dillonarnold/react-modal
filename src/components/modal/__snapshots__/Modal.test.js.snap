// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Modal component should render properly 1`] = `
<Modal
  isShown={true}
  onClose={[Function]}
>
  <div
    className="modal-background modal--hide"
    onClick={[Function]}
  >
    <div
      className="modal"
      role="dialog"
    >
      <div
        className="content"
      />
      <ModalFooter
        closeButtonRef={
          Object {
            "current": <button
              class="footer_close"
              type="button"
          >
              Close
          </button>,
          }
        }
        onClose={[Function]}
      >
        <footer
          className="footer"
        >
          <button
            className="footer_close"
            onClick={[Function]}
            type="button"
          >
            Close
          </button>
        </footer>
      </ModalFooter>
    </div>
  </div>
</Modal>
`;
